tests/reports/unit/test_sonatadata.cpp

```diff
@@ -0,0 +1,127 @@
+                std::initializer_list<nodes_t::value_type>{{1, node}, {2, node2}, {42, node42}});
+
+            int num_steps = 3;
+            size_t max_buffer_size = 1024;
+            std::unique_ptr<SonataData> sonata = std::make_unique<SonataData>(
+                "test_sonatadata", max_buffer_size, num_steps, dt, tstart, tend, nodes);
+            std::vector<uint64_t> nodeids_1 = {1, 42};
+            std::vector<uint64_t> nodeids_2 = {2};
+
+            sonata->prepare_dataset();
+            for (int i = 0; i < num_steps; i++) {
+                sonata->record_data(i, nodeids_1);
+                sonata->record_data(i, nodeids_2);
+                sonata->update_timestep(i);
+            }
+            // sonata->write_data();
+            sonata->close();
+
+            THEN(
+                "The buffer size is the total number of steps times the total number of elements") {
+                // 1024 / sizeof(double) / 11 = 11.6 > 3 (total number of steps)
+                // buffer_size = 11 * 3
+                REQUIRE(sonata->get_report_buffer().size() == 33);
+            }
+
+            THEN("We check the node ids of the sonata report") {
+                const std::vector<uint64_t> node_ids = sonata->get_node_ids();
+                std::vector<uint64_t> compare = {1, 2, 42};
+                REQUIRE(node_ids == compare);
+            }
+
+            THEN("We check the element ids of the sonata report") {
+                const std::vector<uint32_t> element_ids = sonata->get_element_ids();
+                /*std::vector<uint32_t> compare = { 1000, 1001, 2000, 2001, 2002, 42000, 42001,
+                                                  42002, 42003, 42004, 42005 };*/
+                std::vector<uint32_t> compare = {0, 1, 10, 11, 12, 20, 21, 22, 23, 24, 25};
+                REQUIRE(element_ids == compare);
+            }
+
+            THEN("We check the index pointers of the sonata report") {
+                const std::vector<uint64_t> index_pointers = sonata->get_index_pointers();
+                std::vector<uint64_t> compare = {0, 2, 5};
+                REQUIRE(index_pointers == compare);
+            }
+        }
+        WHEN("We record some other data and prepare the dataset for a small max buffer size") {
+            std::shared_ptr<nodes_t> nodes = std::make_shared<nodes_t>(
+                std::initializer_list<nodes_t::value_type>{{1, node}, {2, node2}, {42, node42}});
+            int num_steps = 3;
+            size_t max_buffer_size = 256;
+            std::unique_ptr<SonataData> sonata2 = std::make_unique<SonataData>(
+                "test_sonatadata2", max_buffer_size, num_steps, dt, tstart, tend, nodes);
+
+            sonata2->prepare_dataset();
+            for (int i = 0; i < num_steps; i++) {
+                sonata2->record_data(i);
+                sonata2->update_timestep(i);
+            }
+            // sonata2->write_data();
```

Comment:
Because the data is written inside the update_timestep() method. However It is not really clear so I changed the name to check_and_write() (checks when the buffer is full or all timesteps are recorded and write to file).